File: tools/offgeo/lib/varint.py
"""LEB128-style unsigned varint + zigzag signed-varint encoding, stdlib-
only. Used by `tools/offgeo/prototype-pack-formats.py`'s custom binary
format candidate (`OFF-104`) -- the same encoding style `spec.md` and
`roadmap.md` describe as a real candidate ("improve dictionaries/
varints" is explicitly listed as an R1 size-reduction lever). Kept as
its own small module, not inlined in the prototype script, since a real
R2 compiler/runtime would need the identical encoding on both the
Python (or whatever R1 selects) compiler side and the browser JS reader
side -- this is the reference implementation either would be checked
against.
"""
from __future__ import annotations


def write_uvarint(value: int) -> bytes:
    """Encode a non-negative integer as an unsigned LEB128 varint."""
    if value < 0:
        raise ValueError(f"write_uvarint requires a non-negative value, got {value}")
    out = bytearray()
    while True:
        byte = value & 0x7F
        value >>= 7
        if value:
            out.append(byte | 0x80)
        else:
            out.append(byte)
            return bytes(out)


def read_uvarint(buf: bytes, offset: int) -> tuple[int, int]:
    """Decode one unsigned varint starting at `offset`. Returns
    (value, new_offset)."""
    result = 0
    shift = 0
    while True:
        byte = buf[offset]
        offset += 1
        result |= (byte & 0x7F) << shift
        if not (byte & 0x80):
            return result, offset
        shift += 7


def zigzag_encode(value: int) -> int:
    """Map a signed integer to an unsigned one so small-magnitude
    negatives stay small-varint-encodable (standard protobuf-style
    zigzag: 0,-1,1,-2,2,... -> 0,1,2,3,4,...)."""
    return (value << 1) ^ (value >> 63) if value < 0 else (value << 1)


def zigzag_decode(value: int) -> int:
    return (value >> 1) ^ -(value & 1)


def write_svarint(value: int) -> bytes:
    return write_uvarint(zigzag_encode(value))


def read_svarint(buf: bytes, offset: int) -> tuple[int, int]:
    raw, offset = read_uvarint(buf, offset)
    return zigzag_decode(raw), offset

File: tools/offgeo/prototype-js/varint.mjs
// LEB128-style unsigned varint + zigzag signed-varint encoding, plain
// ES module, no dependencies. JS port of tools/offgeo/lib/varint.py --
// byte layout is required to match exactly (this and the Python module
// must produce/consume identical bytes for prototype-js/packformat.mjs
// to decode blocks tools/offgeo/prototype-benchmark-reader.py exported).
//
// Deliberately uses plain arithmetic (%, Math.floor, *) instead of
// JS's 32-bit bitwise operators (`&`, `>>>`) -- coordinate values here
// stay well within the 32-bit safe range for this county's data, but
// arithmetic is correct for any magnitude up to Number.MAX_SAFE_INTEGER
// without that assumption, matching Python's arbitrary-precision ints
// more closely than a bit-twiddling port would.

export function writeUvarint(value) {
  if (value < 0) {
    throw new RangeError(`writeUvarint requires a non-negative value, got ${value}`);
  }
  const bytes = [];
  let v = value;
  do {
    let byte = v % 128;
    v = Math.floor(v / 128);
    if (v > 0) byte |= 0x80;
    bytes.push(byte);
  } while (v > 0);
  return Uint8Array.from(bytes);
}

/** Returns [value, newOffset]. */
export function readUvarint(buf, offset) {
  let result = 0;
  let multiplier = 1;
  let o = offset;
  for (;;) {
    const byte = buf[o];
    o += 1;
    result += (byte & 0x7f) * multiplier;
    if ((byte & 0x80) === 0) {
      return [result, o];
    }
    multiplier *= 128;
  }
}

export function zigzagEncode(value) {
  return value >= 0 ? value * 2 : -value * 2 - 1;
}

export function zigzagDecode(value) {
  return value % 2 === 0 ? value / 2 : -(value + 1) / 2;
}

export function writeSvarint(value) {
  return writeUvarint(zigzagEncode(value));
}

/** Returns [value, newOffset]. */
export function readSvarint(buf, offset) {
  const [raw, o] = readUvarint(buf, offset);
  return [zigzagDecode(raw), o];
}
